pkg/remoteconfig/repository.go

```diff
@@ -0,0 +1,343 @@
+}
+
+// ConfigState describes an applied config by the agent client.
+type ConfigState struct {
+	Product string
+	ID      string
+	Version uint64
+}
+
+// CachedFile describes a cached file stored by the agent client
+//
+// Note: You may be wondering why this exists when `ConfigState` exists
+// as well. The API for requesting updates does not mandate that a client
+// cache config files. This implementation just happens to do so.
+type CachedFile struct {
+	Path   string
+	Length uint64
+	Hashes map[string][]byte
+}
+
+// An Update contains all the data needed to update a client's remote config repository state
+type Update struct {
+	// TUFRoots contains, in order, updated roots that this repository needs to keep up with TUF validation
+	TUFRoots [][]byte
+	// TUFTargets is the latest TUF Targets file and is used to validate raw config files
+	TUFTargets []byte
+	// TargetFiles stores the raw config files by their full TUF path
+	TargetFiles map[string][]byte
+	// ClientcConfigs is a list of TUF path's corresponding to config files designated for this repository
+	ClientConfigs []string
+}
+
+// Repository is a remote config client used in a downstream process to retrieve
+// remote config updates from an Agent.
+type Repository struct {
+	// TUF related data
+	latestTargets  *data.Targets
+	tufRootsClient *tufRootsClient
+
+	// Config file storage
+	metadata map[string]Metadata
+	configs  map[string]map[string]interface{}
+}
+
+// NewRepository creates a new remote config respository that will track
+// both TUF metadata and raw config files for a client.
+func NewRepository(embeddedRoot []byte) (*Repository, error) {
+	if embeddedRoot == nil {
+		return nil, ErrMalformedEmbeddedRoot
+	}
+
+	configs := make(map[string]map[string]interface{})
+	for _, product := range allProducts {
+		configs[product] = make(map[string]interface{})
+	}
+
+	tufRootsClient, err := newTufRootsClient(embeddedRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Repository{
+		latestTargets:  data.NewTargets(),
+		tufRootsClient: tufRootsClient,
+		metadata:       make(map[string]Metadata),
+		configs:        configs,
+	}, nil
+}
+
+// Update processes the ClientGetConfigsResponse from the Agent and updates the
+// configuration state
+func (r *Repository) Update(update Update) ([]string, error) {
+	// TUF: Update the roots
+	//
+	// NWe don't want to partially update the state, so we need a temporary client to hold the new root
+	// data until we know it's valid
+	tmpRootClient, err := r.tufRootsClient.clone()
+	if err != nil {
+		return nil, err
+	}
+	latestRoot, err := tmpRootClient.updateRoots(update.TUFRoots)
+	if err != nil {
+		return nil, err
+	}
+
+	// 1: Validate and Deserialize the TUF Targets
+	//
+	// Note: This goes further than the RFC requires and validates the TUF targets metadata's signatures.
+	// This is NOT required for most clients per the RFC.
+	updatedTargets, err := unmarshalTargets(latestRoot, update.TUFTargets)
```

Comment:
wouldn't this make more sense as a `validateTargets` method in the tuf partial client ?

Reply:
Ah yes, it would - which would remove the awkward need to have the latest root after a root update in the main update method.